apps/omg_child_chain/lib/omg_child_chain/fees/fee_updater.ex

```diff
@@ -0,0 +1,91 @@
+# Copyright 2019-2020 OmiseGO Pte Ltd
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+defmodule OMG.ChildChain.Fees.FeeUpdater do
+  @moduledoc """
+  Decides whether fees will be updated from the fetched fees from the feed.
+  """
+
+  alias OMG.ChildChain.Fees.FeeMerger
+  alias OMG.Fees
+
+  @type feed_reading_t :: {pos_integer(), Fees.full_fee_t()}
+  @type can_update_result_t :: {:ok, feed_reading_t()} | :no_changes
+
+  @doc """
+  Newly fetched fees will be effective as long as the amount change on any token is significant
+  or the time passed from previous update exceeds the update interval.
+  """
+  @spec can_update(
+          stored_fees :: feed_reading_t(),
+          fetched_fees :: feed_reading_t(),
+          tolerance_percent :: pos_integer(),
+          update_interval_seconds :: pos_integer()
+        ) :: can_update_result_t()
+  def can_update({_, fee_spec}, {_, fee_spec}, _tolerance_percent, _update_interval_seconds), do: :no_changes
+
+  def can_update({t0, _}, {t1, _} = updated, _tolerance_percent, update_interval_seconds)
+      when t0 <= t1 and t1 - t0 >= update_interval_seconds,
+      do: {:ok, updated}
+
+  def can_update({_, stored_fees}, {_, fetched_fees} = updated, tolerance_percent, _update_interval_seconds) do
+    with merged when is_map(merged) <- merge_types(stored_fees, fetched_fees),
+         false <- Enum.any?(merged, &token_mismatch?/1),
+         amount_diffs = Map.values(FeeMerger.merge_specs(stored_fees, fetched_fees)),
+         false <- is_change_significant?(amount_diffs, tolerance_percent) do
+      :no_changes
+    else
+      _ -> {:ok, updated}
+    end
+  end
+
+  # Change is significant when
```

Comment:
it looks like functions below are complicated enough that they could be tested separately. meaning ... extracting them into a separate module where they're public and cover them with tests?

Reply:
I refactored the module heavily - hope it helps